### PR TITLE
Assignment6

### PR DESCRIPTION
**Pull request by dnisyndrum**

### CODE REVIEW

File: Assignment6/Array.cs [hunk 1]
```diff
@@ -1,10 +1,91 @@
 ﻿using System;
+using System.Collections;
 using System.Collections.Generic;
 
+#pragma warning disable CS8653
+#pragma warning disable CS8602
+#pragma warning disable CA2208
+#pragma warning disable CA1303
```
Comment (Fervidusletum) [inline Assignment6/Array.cs]: Don't forget to justify in comments why you're disabling warnings.

File: Assignment6/Array.cs [hunk 2]
```diff
@@ -1,10 +1,91 @@
...
+#pragma warning disable CA2208
+#pragma warning disable CA1303
+
 namespace Assignment6
 {
-    public class Array<T>
+    public class ArrayCollection<T> : ICollection<T>
     {
-        
+        public int Count => _Data.Count;
```
Comment (Fervidusletum) [inline Assignment6/Array.cs]:
```suggestion
        public int Count { get => _Data.Count; }
```
Read-only properties will prevent users of this class from modifying this value, only allowing retrieval.  This also  grabs the current value of `_Data.Count` when the property is requested, whereas the current `public int Count => _Data.Count;` only gets the value of `_Data.Count` once, at instantiation.

File: Assignment6/Array.cs [hunk 3]
```diff
@@ -1,10 +1,91 @@
...
+        public bool IsReadOnly => false;
```
Comment (Fervidusletum) [inline Assignment6/Array.cs]:
```suggestion
        public bool IsReadOnly { get => false; }
```

File: Assignment6/Array.cs [hunk 4]
```diff
@@ -1,10 +1,91 @@
...
+        }
+
+        public bool Contains(T item)
+        {
+            return _Data.Contains(item);
+        }
+
+        public void CopyTo(T[] array, int arrayIndex)
+        {
+            if(arrayIndex < 0 || arrayIndex > Capacity)
```
Comment (Fervidusletum) [inline Assignment6/Array.cs]:
```suggestion
            if(arrayIndex < 0 || arrayIndex >= Capacity)
```
If `arrayIndex == Capacity` then you're already past the last element, which is at `Capacity-1`.

File: Assignment6Tests/ArrayTests.cs [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            Assert.IsTrue(intCol.Contains(1));
+            Assert.IsFalse(intCol.Contains(4));
+        }
+
+        [TestMethod]
+        public void Array_CopyTo_Qapla()
+        {
+            ArrayCollection<int> intCol = new ArrayCollection<int>(4);
+            int[] intArray = new int[4];
+            intCol.CopyTo(intArray, 4);
```
Comment (Fervidusletum) [inline Assignment6Tests/ArrayTests.cs]: Needs an `Assert`.

File: Assignment6/Array.cs [hunk 6]
```diff
@@ -1,10 +1,91 @@
...
+            }
+            Capacity = cap;
+            _Data = new List<T>(cap);
+        }
+
+        public ArrayCollection(ICollection<T> collection)
+        {
+            if (collection is null)
+            {
+                throw new ArgumentNullException("Collection is null", nameof(collection));
```
Comment (Fervidusletum) [inline Assignment6/Array.cs]:
```suggestion
                throw new ArgumentNullException(nameof(collection), "Collection is null");
```
Wrong order for params.

File: Assignment6/Array.cs [hunk 7]
```diff
@@ -1,10 +1,91 @@
...
+        public bool Contains(T item)
+        {
+            return _Data.Contains(item);
+        }
+
+        public void CopyTo(T[] array, int arrayIndex)
+        {
+            if(arrayIndex < 0 || arrayIndex > Capacity)
+            {
+                throw new ArgumentOutOfRangeException("Index is out of bounds of this array.", nameof(arrayIndex));
```
Comment (Fervidusletum) [inline Assignment6/Array.cs]:
```suggestion
                throw new ArgumentOutOfRangeException(nameof(arrayIndex), "Index is out of bounds of this array.");
```

File: Assignment6/Array.cs [hunk 8]
```diff
@@ -1,10 +1,91 @@
...
+            {
+                _Data.Add(item);
+            }
+        }
+
+        public void Add(T item)
+        {
+            if(Capacity == Count)
+            {
+                throw new ArgumentException("Array at max capacity.", nameof(item));
```
Comment (Fervidusletum) [inline Assignment6/Array.cs]: These arguments are in correct order, no change needed. `ArgumentException` is the oddball exception with reversed argument ordering.

File: Assignment6/Array.cs [hunk 9]
```diff
@@ -1,10 +1,91 @@
 ﻿using System;
+using System.Collections;
 using System.Collections.Generic;
 
+#pragma warning disable CS8653
```
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: Why are you disabling this warning for the entire file.  This does not seem appropriate.  The warning is generally valid.

File: Assignment6/Array.cs [hunk 10]
```diff
@@ -1,10 +1,91 @@
 ﻿using System;
+using System.Collections;
 using System.Collections.Generic;
 
+#pragma warning disable CS8653
+#pragma warning disable CS8602
```
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: Why are you disabling this warning for the entire file.  This does not seem appropriate.  The warning is generally valid.

File: Assignment6/Array.cs [hunk 11]
```diff
@@ -1,10 +1,91 @@
 ﻿using System;
+using System.Collections;
 using System.Collections.Generic;
 
+#pragma warning disable CS8653
+#pragma warning disable CS8602
+#pragma warning disable CA2208
```
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: Why are you disabling this warning for the entire file.  This does not seem appropriate.  The warning is generally valid.

File: Assignment6/Array.cs (hunk 1 above)
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: I agree with disabling this waring as we are not trying to be localizable.

File: Assignment6/Array.cs [hunk 12]
```diff
@@ -1,10 +1,91 @@
 ﻿using System;
```
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: You should change the file name to match the class name.

File: Assignment6/Array.cs [hunk 13]
```diff
@@ -1,10 +1,91 @@
...
+        public void CopyTo(T[] array, int arrayIndex)
+        {
+            if (arrayIndex < 0 || arrayIndex >= Capacity)
+            {
+                throw new ArgumentOutOfRangeException(nameof(arrayIndex), "Index is out of bounds of this array.");
+            }
+            _Data.CopyTo(array, arrayIndex);
+        }
+
+        public IEnumerator<T> GetEnumerator()
```
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: Consider using an expression bodied member:
```
public IEnumerator<T> GetEnumerator() => _Data.GetEnumerator();
```

File: Assignment6/Array.cs [hunk 14]
```diff
@@ -1,10 +1,91 @@
 ﻿using System;
+using System.Collections;
 using System.Collections.Generic;
```
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: All disabling should provide an explanation and an indication of what the warning is that is getting disabled.

File: Assignment6/Array.cs (hunk 13 above)
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: Consider implementing as an expression bodied member:
```
        public IEnumerator<T> GetEnumerator() =>_Data.GetEnumerator();
```

File: Assignment6Tests/Assignment6Tests.csproj [hunk 15]
```diff
@@ -0,0 +1,21 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (MarkMichaelis) [inline Assignment6Tests/Assignment6Tests.csproj]: Rename file to Assignment6.Tests.csproj

**Comment by MarkMichaelis** [comment]
##  (dnisyndrum)

## Assignment6

| Build/Compile (Build Succeeded: 0 Error(s), 0 Warning(s))| **10** |
| --------------------- | ------ |
| Assignment compiles even though major pieces might be missing, or there are significant problems: **✔** |  **5** |
| The assignment contains only a few required elements or has significant problems. | **5** |


| Assignment  | **15** |
| --------------------- | ------ |
| Supports collecting a fixed number of items once instantiated. | ✔ |
| Has a constructor via which you can specify fixed width.  (Other constructors are welcome.) | ✔ |
| Handle null appropriately when items are missing (this is likely a judgement call so justify your answer in the comments.) | ✔ |
| Throw an exception if an item doesn't exist. | ✔ |
| Include an index operator. | ✔ |
| Support using foreach over the items in the `Array<T>` class. | ✔ |
| Support a collection initializer. | ✔ |
| Implement `System.Collections.Generic.ICollection<T>`. | ✔ |
| Ignore support for multi-dimensional arrays. | ✔ |
| Add a `Capacity` property to return the capacity (the value specified in the constructor) of the collection. | ✔ |


| Project Analysis |   |
| -----------------| - |
| Nullable Enabled: | **✔** |
| CodeReview Performed | **✔** |
| Add the code analysis packages and appropriately handle all warnings* |  **❌** |
*<sub><sup>`IntelliTectAnalyzer.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.CodeQuality.Analyzers.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.NetFramework.Analyzers.dll`. (It is okay to appropriately disable rules such as  localization/culture rules.)</sub></sup>


| ProjectName | Nullable | Solutions | Analyzers | Unexpected Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| Assignment6 | Enable | Assignment6.sln |  | CA2208, CS8602, CS8653 | 4 |
| Assignment6Tests | Enable | Assignment6.sln |  | CS1701, CS1702 | 4 |


| Testing | **20** |
| --------------------- | ------ |
| Test Run Succeeded: Total tests=18, Passed=18, Failed=0, Skipped=0 | **95%** |
| Unit test project fails to compile, most test fails, or contains significant problems. | **10** |
| Most assigned work is tested or contain some errors. | **5** |
| All assigned work is tested | **5** |


| Extra Credit | **<0 - 10>** |
| --------------------- | ------ |
| Write a cast operator to a C# array or a `System.Collections.Generic.List<T>`. | ❌ |
| Add support for an `IReadOnlyArray` interface (for you to define) that provides readonly access to the count and each element in the array.  Be sure to appropriately decorate it for covariance and contravariance | ❌ |

**Pull request closed without merging**